src/hotspot/share/logging/logDecorations.cpp

```diff
@@ -53,89 +48,93 @@ const char* LogDecorations::host_name() {
-  ASSERT_AND_RETURN(written, pos)
+void LogDecorations::print_uptime_decoration(outputStream* st) const {
+  st->print("%.3fs", _elapsed_seconds);
 }
 
-// Small helper for uptime conversion
-static jlong elapsed_time(int unit_multiplier) {
-  return (jlong)(os::elapsedTime() * unit_multiplier);
+void LogDecorations::print_timemillis_decoration(outputStream* st) const {
+  st->print(INT64_FORMAT "ms", (int64_t)_millis);
 }
 
-char * LogDecorations::create_uptimemillis_decoration(char* pos) {
-  int written = jio_snprintf(pos, DecorationsBufferSize - (pos - _decorations_buffer),
-                             INT64_FORMAT "ms", elapsed_time(MILLIUNITS));
-  ASSERT_AND_RETURN(written, pos)
+void LogDecorations::print_uptimemillis_decoration(outputStream* st) const {
+  st->print(INT64_FORMAT "ms", (int64_t)(_elapsed_seconds * MILLIUNITS));
 }
 
-char * LogDecorations::create_timenanos_decoration(char* pos) {
-  int written = jio_snprintf(pos, DecorationsBufferSize - (pos - _decorations_buffer), INT64_FORMAT "ns", os::javaTimeNanos());
-  ASSERT_AND_RETURN(written, pos)
+void LogDecorations::print_timenanos_decoration(outputStream* st) const {
+  st->print(INT64_FORMAT "ns", (int64_t)_nanos);
 }
 
-char * LogDecorations::create_uptimenanos_decoration(char* pos) {
-  int written = jio_snprintf(pos, DecorationsBufferSize - (pos - _decorations_buffer), INT64_FORMAT "ns", elapsed_time(NANOUNITS));
-  ASSERT_AND_RETURN(written, pos)
+void LogDecorations::print_uptimenanos_decoration(outputStream* st) const {
+  st->print(INT64_FORMAT "ns", (int64_t)(_elapsed_seconds * NANOUNITS));
 }
 
-char * LogDecorations::create_pid_decoration(char* pos) {
-  int written = jio_snprintf(pos, DecorationsBufferSize - (pos - _decorations_buffer), "%d", os::current_process_id());
-  ASSERT_AND_RETURN(written, pos)
+void LogDecorations::print_pid_decoration(outputStream* st) const {
+  st->print("%d", _pid);
 }
 
-char * LogDecorations::create_tid_decoration(char* pos) {
-  int written = jio_snprintf(pos, DecorationsBufferSize - (pos - _decorations_buffer),
-                             INTX_FORMAT, os::current_thread_id());
-  ASSERT_AND_RETURN(written, pos)
+void LogDecorations::print_tid_decoration(outputStream* st) const {
+  st->print(INTX_FORMAT, _tid);
 }
 
-char* LogDecorations::create_level_decoration(char* pos) {
-  // Avoid generating the level decoration because it may change.
-  // The decoration() method has a special case for level decorations.
-  return pos;
+void LogDecorations::print_level_decoration(outputStream* st) const {
+  st->print_raw(LogLevel::name(_level));
 }
 
-char* LogDecorations::create_tags_decoration(char* pos) {
-  int written = _tagset.label(pos, DecorationsBufferSize - (pos - _decorations_buffer));
-  ASSERT_AND_RETURN(written, pos)
+void LogDecorations::print_tags_decoration(outputStream* st) const {
+  char tmp[255];
+  _tagset.label(tmp, sizeof(tmp)); // Todo: add a LogTagSet::label(outputStream*) to avoid this copy step
```

Comment:
Yes, that would be desirable and probably not too hard :)

Otherwise you'll might have an uninitialized `tmp` array if the first tag doesn't fir into the array and  `LogTagSet::label()` returns `-1`. I understand that that's more a theoretical problem because at least one tag should always fir within 255 characters :)

Reply:
My fist version had this but Xin mentioned the patch was too large :) I'll reinstate my first solution.